at_client/lib/src/service/sync_service_impl.dart

```diff
@@ -45,76 +53,117 @@ class SyncServiceImpl implements SyncService, AtSignChangeListener {
-          'AT0014', 'Sync-InProgress. Cannot start a new sync process');
-      if (onError != null) {
-        onError(syncResult);
+  void sync({Function? onDone}) {
+    final syncRequest = SyncRequest();
+    syncRequest.onDone = onDone;
+    syncRequest.requestedOn = DateTime.now().toUtc();
+    _addSyncRequestToQueue(syncRequest);
+    return;
+  }
+
+  /// Listens on stats notification sent by the cloud secondary server
+  Future<void> _statsServiceListener() async {
+    _statsNotificationListener = await NotificationServiceImpl.create(_atClient)
+        as NotificationServiceImpl;
+    // Setting the regex to 'statsNotification' to receive only the notifications
+    // from stats notification service.
+    _statsNotificationListener
+        .subscribe(regex: 'statsNotification')
+        .listen((notification) async {
+      final serverCommitId = notification.value;
+      if (serverCommitId != null &&
+          int.parse(serverCommitId) > await _getLocalCommitId()) {
+        final syncRequest = SyncRequest();
+        syncRequest.onDone = _onDone;
+        syncRequest.onError = _onError;
+        syncRequest.requestSource = SyncRequestSource.system;
+        syncRequest.requestedOn = DateTime.now().toUtc();
+        _addSyncRequestToQueue(syncRequest);
       }
-      return syncResult;
+    });
+  }
+
+  void _processSyncRequests() async {
+    if (!await NetworkUtil.isNetworkAvailable()) {
+      _logger.finer('skipping sync due to network unavailability');
+    }
+    if (_syncRequests.length < _syncRequestThreshold &&
+        (_syncRequests.isNotEmpty &&
+            _syncRequests
+                    .elementAt(0)
+                    .requestedOn
+                    .difference(DateTime.now().toUtc())
+                    .inSeconds <
+                _syncRequestTriggerInSeconds)) {
+      _logger.finer(
+          'skipping sync - queue length ${_syncRequests.length} - first request time ${_syncRequests.elementAt(0).requestedOn}');
     }
+    final syncRequest = _syncRequests.removeFirst();
     try {
-      //Setting isSyncInProgress to true to prevent parallel sync calls.
-      _isSyncInProgress = true;
-      // If network is not available, return.
-      if (!await NetworkUtil.isNetworkAvailable()) {
-        _logger.severe('Failed connecting to internet');
-        syncResult = SyncResult();
-        syncResult.syncStatus = SyncStatus.failure;
-        syncResult.atClientException =
-            AtClientException('AT0014', 'Failed connecting to internet');
-        _isSyncInProgress = false;
-        if (onError != null) {
-          onError(syncResult);
-        }
-        return syncResult;
+      if (await isInSync()) {
+        _logger.finer('server and local are in sync - ${syncRequest._id}');
+        syncRequest.result!
+          ..syncStatus = SyncStatus.success
+          ..lastSyncedOn = DateTime.now().toUtc();
+        _syncComplete(syncRequest);
```

Comment:
We have to call this on one of the app sync requests. The first one need not be an app sync request. If all of them are system sync requests then it is ok. We have to track the last app sync request seperately or something.

Reply:
will tackle this after initial round of testing